### PR TITLE
Static pipelines now handle types with PipelineColumn properties.

### LINKED ISSUE #1085: Expand PiGSTy's schema shape type support
PiGSTy's (#632) schema propagation system relies on a static type that is currently restricted to be a value-tuple. There is, however, no fundamental reason for it to be value-tuples, or only value-tuples; my motivation was merely just that if we consider the C# syntax for declaring value-tuples vs. anonymous types:

```csharp
var x = (a: 1, b: 2);
var y = new { A = 1, B = 2 };
```

one clearly had more elegant syntax than the other. Yet there are reasons to *also* support classes with `PipelineColumn` properties. (This includes but is not limited to anonymous classes).

1. Intellisense for value-tuples, especially in the context of generics and especially *especially* in the context of delegates, is currently not the best. [This may be addressed, or it may not.](https://github.com/dotnet/roslyn/issues/29949) At any rate even if addressed we would want our API to have a good experience in older versions of Visual Studio.

2. My possibly imperfect understanding from @dsyme is that F# can work with anonymous types gracefully, but as far as I am aware cannot yet work with value-tuples at all. Considering that the intersection of F# developers and ML.NET potential users is high at least when compared to the general population of .NET developer, this may be viewed as a priority.

3. For explicitly defined (i.e., non-anonymous) classes, it allows one to define something like `DataView<MyData>`, which has the advantage of being easily expressable as a return value and a parameter to a function. Could someone write something like `DataView<(Scalar<bool> label, Vector<float> features)>` as a return value or parameter? Yes, absolutely, I suppose, but it's not very pretty. Also one thing I learnt to my surprise is that, at least as of C# 7.3, one *cannot* merely do something like `using Foo = (int bar, float biz)`, so even a shorthand based on type aliases even within a single file is not possible.

Currently the set of allowed "schema shape" types is the following:

1. Any of the public subclasses of `PipelineColumn`.
2. A value-tuple whose items are of an allowed "schema shape" type.

What I am proposing is that we expand this definition to include:

3. Classes whose only publicly accessible members are properties that are one of the basic abstract subtypes of `PipelineColumn` and a single public constructor in one of the following two subtypes:

    * The properties are get-only, and the single constructor has parameters where a one-to-one correspondence with each of the properties, where the parameter and property have identical types, and this constructor sets the properties, and no other work.

    * The properties have get-set accessors, and the single construct has no arguments. (Note that this can be a default constructor.)

    In each case, whether the property is set via constructor parameter or just by assigning it, the property should always be exactly the object that was either passed in or assigned.

Note that due to the way the validation code works, and for several practical reasons, it is best and easiest to allow a mix of these types -- that is, just as you can have nested value-tuples, you could have, for example, a value-tuple nested in a class, nested in a value-tuple.

I am considering the wisdom of allowing fields. On the one hand, we allow fields on the classes for `PredictionEngine`. These of course cannot be the same object (one describes values of a column, whereas one describes the properties of the columns themselves and how they are meant to be configured), but it may be that allowing fields in one but not the other may lead to confusion. Not sure about this. For now I am thinking to only allow properties, and if someone complains we can consider adding it then.

## Work

We might identify five things that should be done. The first three should probably be in one PR. The fourth can happen in a subsequent PR. The fifth is sort of a "nice to have" but strictly speaking unnecessary. One of those things to be done in someone's copious spare time.

1. The methods of `StaticPipeInternalUtils` should be expanded to also accommodate the aforementioned types.

2. The code analyzer that checks the type parameters with `IsShapeAttribute` should be similarly expanded.

3. In several places throughout the code base the generic type parameter has `Tuple` in its name, e.g., `TTupleShape`, `TTupleInShape`, and so forth. These ought to be changed to things like `TShape` or `TInShape` and the like to make them less specific to tuples.

4. Part of the reasons why we are doing this is to make the structure work better in F#. F# support for value tuples is limited. This suggests that those so-called "pigstentions" where we return a set of columns (e.g., binary classification), where we currently have a value-tuple (e.g., `(Scalar<float> score, Scalar<float> probability, Scalar<bool> predictedLabel)`) we may be better served by instead having a definite type, `BinaryClassificationOutput` looking something like this:

   ```csharp
   public sealed BinaryClassificationOutput
   {
       public Scalar<float> Score { get; }
       public Scalar<float> Probability { get; }
       public Scalar<bool> PredictedLabel { get; }

       public BinaryClassificationOutput(...) { ... }
   }
   ```

5. Having a user responsible for defining the types is fairly obnoxious in its own right. So, we might imagine a code action through a Roslyn code-fix provider that can take an `Estimator<T>` where `T` is some sort of inline type (either value-tuple or anonymous class), based on that class defines an explicit type (by default perhaps some `private` nested class), and then uses it to provide `Estimator<T2>` over this new hypothetical type `T2`. This would be relatively easy for the user to call, and then they could modify it as they like. (E.g., make it public, move it somewhere else, add XML docs, and so forth.)

### PR DESCRIPTION
**Pull request by TomFinley**
* Update the internal infrastructure to accomodate these types,
* Update the Roslyn analyzer to accomodate these types.
* Update the tests so that they exercise this capability.
* Opportunistically fix some problems with the Roslyn analyzer
  brought up in this work.

Fixes #1085.

### CODE REVIEW

**Comment by TomFinley** [comment]
Oh macOS Build_Debug... you kidder.

File: src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs [hunk 1]
```diff
@@ -37,28 +37,28 @@ public static class StaticPipeUtils
         /// <typeparam name="TIn">The type parameter for the input type to the data reader estimator.</typeparam>
         /// <typeparam name="TDelegateInput">The input type of the input delegate. This might be some object out of
         /// which one can fetch or else retrieve </typeparam>
-        /// <typeparam name="TTupleOutShape"></typeparam>
+        /// <typeparam name="TOutShape"></typeparam>
```
Comment (Zruty0) [inline src/Microsoft.ML.Data/StaticPipe/StaticPipeUtils.cs]: >typeparam [](start = 13, length = 9)

remove empty

**Comment by Zruty0** [comment]
> namespace Microsoft.ML.StaticPipe

maybe rename to `LightGbmStatic` while you're at it?..

---
Refers to: src/Microsoft.ML.LightGBM/LightGBMStatics.cs:12 in 22399ab. [](commit_id = 22399abc050df8782d9d170e57f3f4b5c52743d7, deletion_comment = False)

**Comment by Zruty0** [comment]
> namespace Microsoft.ML.StaticPipe

`Statics` again. Where were I looking?..

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs:12 in 22399ab. [](commit_id = 22399abc050df8782d9d170e57f3f4b5c52743d7, deletion_comment = False)

File: test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs [hunk 2]
```diff
@@ -126,6 +126,76 @@ void CheckValuesSame(bool bl, string tx, float v0, float v1, float v2)
...
+
+            var schema = text.AsDynamic.GetOutputSchema();
+            Helper(schema, "yo.Foo", TextType.Instance);
+            Helper(schema, "yo.Bar", new VectorType(NumberType.Float, 5));
+            Helper(schema, "dawg.Biz", TextType.Instance);
+            Helper(schema, "dawg.Blam", new VectorType(NumberType.R8, 2));
+
+            Helper(schema, "how.Stuff.hi", BoolType.Instance);
+            Helper(schema, "how.Stuff.my.Foo", TextType.Instance);
+            Helper(schema, "how.Stuff.my.Bar", new VectorType(NumberType.Float, 4));
```
Comment (Zruty0) [inline test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs]: >how.Stuff.my.Bar [](start = 28, length = 16)

'how stuff my bar', hmm...

In the spirit of open-source, should we have a bit more, um, mature naming scheme?..

Comment (TomFinley) [reply test/Microsoft.ML.StaticPipelineTesting/StaticPipeTests.cs]: I'll change `Stuff` to `Donut`?

---
In reply to: [222036765](https://github.com/dotnet/machinelearning/pull/1115#discussion_r222036765) [](ancestors = 222036765)

**Comment by TomFinley** [comment]
> namespace Microsoft.ML.StaticPipe

Sounds good sure.

---
In reply to: [426354064](https://github.com/dotnet/machinelearning/pull/1115#issuecomment-426354064) [](ancestors = 426354064)

---
Refers to: src/Microsoft.ML.LightGBM/LightGBMStatics.cs:12 in 22399ab. [](commit_id = 22399abc050df8782d9d170e57f3f4b5c52743d7, deletion_comment = False)

**Comment by TomFinley** [comment]
> namespace Microsoft.ML.StaticPipe

Renamed

---
In reply to: [426354251](https://github.com/dotnet/machinelearning/pull/1115#issuecomment-426354251) [](ancestors = 426354251)

---
Refers to: src/Microsoft.ML.StandardLearners/Standard/LogisticRegression/LbfgsStatics.cs:12 in 22399ab. [](commit_id = 22399abc050df8782d9d170e57f3f4b5c52743d7, deletion_comment = False)

**Comment by TomFinley** [comment]
Same test failed... MulticlassTreeFeaturizedLRTest. But with a different bad range. Not sure what's up with that, I definitely didn't touch anything related to that test.

---
In reply to: [426148269](https://github.com/dotnet/machinelearning/pull/1115#issuecomment-426148269) [](ancestors = 426148269)

**Pull request merged**